client/java/src/main/java/com/linecorp/centraldogma/internal/client/ReplicationLagTolerantCentralDogma.java

```diff
@@ -0,0 +1,485 @@
+                                   });
+                });
+    }
+
+    @Override
+    public <T> CompletableFuture<Entry<T>> watchFile(
+            String projectName, String repositoryName, Revision lastKnownRevision,
+            Query<T> query, long timeoutMillis) {
+
+        return normalizeRevisionAndExecute(
+                projectName, repositoryName, lastKnownRevision,
+                normLastKnownRevision -> {
+                    return delegate.watchFile(projectName, repositoryName, normLastKnownRevision,
+                                              query, timeoutMillis)
+                                   .thenApply(entry -> {
+                                       if (entry != null) {
+                                           updateLatestKnownRevision(projectName, repositoryName,
+                                                                     entry.revision());
+                                       }
+                                       return entry;
+                                   });
+                });
+    }
+
+    private <T> CompletableFuture<T> normalizeRevisionAndExecute(
+            String projectName, String repositoryName, Revision revision,
+            Function<Revision, CompletableFuture<T>> taskRunner) {
+        return normalizeRevision(projectName, repositoryName, revision)
+                .thenCompose(normRev -> execute(
+                        () -> taskRunner.apply(normRev),
+                        (res, cause) -> cause != null &&
+                                        handleRevisionNotFound(projectName, repositoryName, normRev, cause)));
+    }
+
+    private <T> CompletableFuture<T> normalizeRevisionsAndExecute(
+            String projectName, String repositoryName, Revision from, Revision to,
+            BiFunction<Revision, Revision, CompletableFuture<T>> taskRunner) {
+
+        return CompletableFutures.allAsList(ImmutableList.of(
+                normalizeRevision(projectName, repositoryName, from),
+                normalizeRevision(projectName, repositoryName, to)))
+                .thenCompose(normRevs -> {
+                    final Revision normFromRev = normRevs.get(0);
+                    final Revision normToRev = normRevs.get(1);
+                    return execute(
+                            () -> taskRunner.apply(normFromRev, normToRev),
+                            (res, cause) -> {
+                                if (cause == null) {
+                                    return false;
+                                }
+                                final Revision normRev = normFromRev.compareTo(normToRev) > 0 ? normFromRev
+                                                                                              : normToRev;
+                                return handleRevisionNotFound(projectName, repositoryName, normRev, cause);
+                            });
+                });
+    }
+
+    private <T> CompletableFuture<T> execute(
```

Comment:
nit: Even though this is a private method, in my case, I'm a little confused on this method name.
How about renaming `execute(...)` to `excuteWithRetry(...)`?

Reply:
Renamed to `executeWithRetries()` and added some Javadoc for clarity.